exposed-dao/src/main/kotlin/org/jetbrains/exposed/dao/EntityCache.kt

```diff
@@ -42,29 +46,31 @@ class EntityCache(private val transaction: Transaction) {
     }
 
     fun <ID:Comparable<ID>, T: Entity<ID>> scheduleInsert(f: EntityClass<ID, T>, o: T) {
-        inserts.getOrPut(f.table) { arrayListOf() }.add(o as Entity<*>)
+        inserts.getOrPut(f.table) { LinkedIdentityHashSet() }.add(o as Entity<*>)
+    }
+
+    fun <ID:Comparable<ID>, T: Entity<ID>> scheduleUpdate(f: EntityClass<ID, T>, o: T) {
+        updates.getOrPut(f.table) { LinkedIdentityHashSet() }.add(o as Entity<*>)
```

Comment:
The array list previously used required O(N) to determine if the entity needs insertion.  That means that every bulk flush operation was O(n^2), since it would check `isNewEntity` on every entity.  Yikes!  I've maintained a list of updates using a `LinkedIdentityHashSet`, which maintains order, is O(1) for contains, and uses the identity hash code so no ID is required.

Reply:
@Tapac does the order of updates matter?  It clearly does for inserts, but it is hard to tell for updates.